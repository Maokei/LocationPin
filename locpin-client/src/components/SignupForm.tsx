import React, { useState, useEffect } from "react";
import axios from "axios";
import { RouteComponentProps, withRouter } from 'react-router-dom';
import { TextField } from "@material-ui/core";
import Grid from "@material-ui/core/Grid";
import MailOutlineIcon from "@material-ui/icons/MailOutline";
import HowToRegIcon from "@material-ui/icons/HowToReg";
import AccountCircle from "@material-ui/icons/AccountCircle";
import VpnKeyIcon from "@material-ui/icons/VpnKey";
import Button from "@material-ui/core/Button";

const SignupForm = (): JSX.Element => {
    const [email, setEmail] = useState("");
    const [name, setName] = useState("");
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [isButtonDisabled, setIsButtonDisabled] = useState(true);
    const [helperText, setHelperText] = useState("");
    let errorArray:string[] = [];

    const handleSubmit = async () => {
        const data = {
            email,
            name,
            password,
            username,
        };

        const emailRegex = /^([A-Za-z0-9_\-.+])+@([A-Za-z0-9_\-.])+\.([A-Za-z]{2,})$/; 
        if (emailRegex.test(email) && (email.length > 2) && (41 < email.length)) {
            errorArray.push("Bad email");
        }
        if((name.length > 2) && (41 < name.length)) {
            errorArray.push("Name needs to between 3 and 40 characters.");
        }
        if((username.length > 2) && (16 < username.length)) {
            errorArray.push("Name needs to between 3 and 15 characters.");
        }
        if((password.length > 2) && (101 < password.length)) {
            errorArray.push("Password needs to between 3 and 100 characters.");
        }
        
        if(errorArray.length > 0) {
            return;
        }
        
        const res: any = await axios.post(
            "http://localhost:8080/api/auth/signup",
            data
        );
        errorArray = [];
        console.log({data});
    };

    useEffect(() => {
        if (email.trim() && name.trim() && username.trim() && password.trim()) {
            setIsButtonDisabled(false);
        } else {
            setIsButtonDisabled(true);
        }
    }, [email, name, username, password]);

    return (
        <form
            style={{ padding: "0 1em" }}
            className="center"
            noValidate
            autoComplete="off">
            <Grid container spacing={1} alignItems="flex-end">
                <Grid item>
                    <MailOutlineIcon />
                </Grid>
                <Grid item>
                    <TextField
                        id="input-with-icon-grid"
                        className="input-field"
                        label="Email"
                        name="email"
                        type="email"
                        error={(errorArray.length > 0)}
                        onChange={(e) => setEmail(e.target.value)}
                    />
                </Grid>
            </Grid>
            <br />
            <Grid container spacing={1} alignItems="flex-end">
                <Grid item>
                    <HowToRegIcon />
                </Grid>
                <Grid item>
                    <TextField
                        id="input-with-icon-grid"
                        className="input-field"
                        label="Name"
                        error={(errorArray.length > 0)}
                        onChange={(e) => setName(e.target.value)}
                    />
                </Grid>
            </Grid>
            <br />
            <Grid container spacing={1} alignItems="flex-end">
                <Grid item>
                    <AccountCircle />
                </Grid>
                <Grid item>
                    <TextField
                        id="input-with-icon-grid"
                        className="input-field"
                        label="Username"
                        error={(errorArray.length > 0)}
                        onChange={(e) => setUsername(e.target.value)}
                    />
                </Grid>
            </Grid>
            <br />
            <Grid container spacing={1} alignItems="flex-end">
                <Grid item>
                    <VpnKeyIcon />
                </Grid>
                <Grid item>
                    <TextField
                        id="input-with-icon-grid"
                        className="input-field"
                        label="Password"
                        type="password"
                        error={(errorArray.length > 0)}
                        helperText={helperText}
                        onChange={(e) => setPassword(e.target.value)}
                    />
                </Grid>
            </Grid>
            <br />

            <Button
                className="btn"
                variant="outlined"
                onClick={handleSubmit}
                disabled={isButtonDisabled}>
                Submit
            </Button>
        </form>
    );
};

export default SignupForm;