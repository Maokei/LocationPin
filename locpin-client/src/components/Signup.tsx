import React from "react";
import Grid from "@material-ui/core/Grid";
import { Container } from "@material-ui/core";
import { PinLogo } from "./PinLogo";
import SignupForm from "./SignupForm";
import Link from "@material-ui/core/Link";

export const Signup = () => {
    return (
        <Container>
            <PinLogo />
            <h2 className="center">Sign Up</h2>
            <SignupForm />
            <Grid container className="link-center">
                {/* <Link href="#">
                    Fogot password?
                </Link> */}
                <p>
                    Already got an account?
                    <Link href="/">Log In</Link>
                </p>
            </Grid>
        </Container>
    );
};
