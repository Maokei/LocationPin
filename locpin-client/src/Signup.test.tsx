import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Signup from "./components/Signup";
import SignupForm from "./components/SignupForm";

Enzyme.configure({ adapter: new Adapter() });

describe("<Signup />", () => {
    const wrapper = shallow(<Signup />);
    // console.log(wrapper.debug());

    it("Signup matches snapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("signup page has a title component", () => {
        const title = wrapper.find("h2").at(1);
        expect(title).toBeTruthy();
    });

    it("Form can be found in Signup", () => {
        const form = wrapper.find("form").at(2);
        expect(form).toBeTruthy();
    });
});

describe("SignupForm", () => {
    let wrapper: Enzyme.ReactWrapper<
        any,
        Readonly<{}>,
        React.Component<{}, {}, any>
    >;
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation((init: any) => {
        return [init, setState];
    });

    beforeEach(() => {
        wrapper = mount(shallow(<SignupForm />).get(0));
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    const formWrapper = mount(<SignupForm />);
    const form = formWrapper.find("form").at(0);
    // console.log(form.debug());
});
