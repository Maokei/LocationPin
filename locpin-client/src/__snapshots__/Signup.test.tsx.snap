// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> Signup matches snapshot 1`] = `ShallowWrapper {}`;
